# Analysis of the results

Here we analyse the utility results of the StarGANv2-VC. Three experiments were run with different seeds.

In [7]:
import os
import numpy as np

In [8]:
# define the dirs that will be included in this analysis
log_dirs = [
    "../../logs/whisper_fastpitch/privacy_0",
    "../../logs/whisper_fastpitch/privacy_1",
    "../../logs/whisper_fastpitch/privacy_2",
    "../../logs/whisper_fastpitch/utility",
]

In [9]:
# map the components to the keys of the files that will be loaded
components = {
    "whisper-small": 4,
    "whisper-large": 4,
    "ser-audeering-w2v": 2,
    "nisqa": 2,
    "asv-plda/ignorant/results": 3,
    "asv-plda/lazy-informed/results": 3,
}

In [10]:
# aggregate results by the keys of the target and source speakers
results = dict()

for component, key in components.items():
    for log_dir in log_dirs:
        file = "eer.txt" if "privacy" in log_dir else "all.txt"
        if not os.path.exists(os.path.join(log_dir, "eval", component)):
            continue
        path = os.path.join(log_dir, "eval", component, file)
        with open(path, "r") as f:
            next(f)  # skip header

            for line in f:
                # get the value for the given dataset and key to the results
                elements = line.split()
                dataset = elements[0]
                value = elements[key]
                if dataset.endswith(".txt"):
                    dataset = dataset[:-4]
                # add the dataset to the results if needed
                if component not in results:
                    results[component] = dict()
                if dataset not in results[component]:
                    results[component][dataset] = list()

                results[component][dataset].append(round(float(value), 4))

In [11]:
# average the results of the different components
avgs = dict()
for component in results:
    avgs[component] = dict()
    for dataset in results[component]:
        avgs[component][dataset] = sum(results[component][dataset]) / len(results[component][dataset])

In [12]:
# print the results in a markdown table
print("| |", end=" ")  # empty cell in the top left corner
# print the dataset names as column headers
for dataset in avgs[list(avgs.keys())[0]]:
    print(dataset, end=" | ")
print()  # new line

# add hyphens to separate the header from the table
print("|", end=" ")
for i in range(len(avgs[list(avgs.keys())[0]]) + 1):
    print("---", end=" | ")
print()  # new line

# print the averages of each dataset for each ASV config
for component in avgs:
    print("|", end=" ")
    print(component, end=" | ")
    for dataset in avgs[component]:
        print(round(avgs[component][dataset], 4), end=" | ")
    print()  # new line


| | cv-test_3utts | ls-test-clean | edacc-test | ravdess | 
| --- | --- | --- | --- | --- | 
| whisper-small | 0.19 | 0.08 | 0.34 | 0.01 | 
| whisper-large | 0.18 | 0.08 | 0.34 | 0.01 | 
| ser-audeering-w2v | 1.0 | 0.99 | 0.99 | 1.0 | 
| nisqa | 3.87 | 3.7 | 3.58 | 3.62 | 
| asv-plda/ignorant/results | 0.4941 | 0.4863 | 0.4843 | 0.4855 | 
| asv-plda/lazy-informed/results | 0.4745 | 0.4316 | 0.4675 | 0.5121 | 


### Utility & Privacy results

| | cv-test_3utts | ls-test-clean | edacc-test | ravdess | 
| --- | --- | --- | --- | --- |
| whisper-small | 0.19 | 0.08 | 0.34 | 0.01 | 
| whisper-large | 0.18 | 0.08 | 0.34 | 0.01 | 
| ser-audeering-w2v | 1.0 | 0.99 | 0.99 | 1.0 | 
| nisqa | 3.87 | 3.7 | 3.58 | 3.62 | 
| asv-plda/ignorant/results | 0.5 | 0.48 | 0.49 | 0.48 | 
| asv-plda/lazy-informed/results | 0.47 | 0.43 | 0.47 | 0.51 | 